cmd/clusterctl/pkg/client/repository/repository_github.go

```diff
@@ -0,0 +1,323 @@
+	}
+
+	// Check if the repository Path is in the expected format.
+	t := strings.Split(strings.TrimPrefix(rURL.Path, "/"), "/")
+	if len(t) < 5 || t[2] != githubReleaseRepository {
+		return nil, errors.Errorf(
+			"invalid url: a GitHub repository url should be in the form https://github.com/{owner}/{Repository}/%s/{latest|version-tag}/{componentsClient.yaml}",
+			githubReleaseRepository,
+		)
+	}
+
+	// extracts all the info from path
+	owner := t[0]
+	repository := t[1]
+	defaultVersion := t[3]
+	path := strings.Join(t[4:], "/")
+
+	// use path's directory as a rootPath
+	rootPath := filepath.Dir(path)
+	// use the file name (if any) as componentsPath
+	componentsPath := getComponentsPath(path, rootPath)
+
+	repo := &gitHubRepository{
+		providerConfig:        providerConfig,
+		configVariablesClient: configVariablesClient,
+		owner:                 owner,
+		repository:            repository,
+		defaultVersion:        defaultVersion,
+		rootPath:              rootPath,
+		componentsPath:        componentsPath,
+	}
+
+	token, err := configVariablesClient.Get(githubTokeVariable)
+	if err != nil {
+		klog.V(1).Infof("The %q configuration variable is missing. Falling back to unauthenticated requests that allows for up to 60 requests per hour.", githubTokeVariable)
+	}
+	if err == nil {
+		repo.setClientToken(token)
+	}
+
+	if defaultVersion == githubLatestReleaseLabel {
+		repo.defaultVersion, err = repo.getLatestRelease()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get GitHub latest version")
+		}
+	}
+
+	return repo, nil
+}
+
+// getComponentsPath returns the file name
+func getComponentsPath(path string, rootPath string) string {
+	// filePath = "/filename"
+	filePath := strings.TrimPrefix(path, rootPath)
+	// componentsPath = "filename"
+	componentsPath := strings.TrimPrefix(filePath, "/")
+	return componentsPath
+}
+
+// getClient returns a github API client
+func (g *gitHubRepository) getClient() *github.Client {
+	if g.injectClient != nil {
+		return g.injectClient
+	}
+	return github.NewClient(g.authenticatingHTTPClient)
+}
+
+// setClientToken sets authenticatingHTTPClient field of gitHubRepository struct
+func (g *gitHubRepository) setClientToken(token string) {
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	g.authenticatingHTTPClient = oauth2.NewClient(context.Background(), ts)
+}
+
+// getVersions returns all the release versions for a github repository
+func (g *gitHubRepository) getVersions() ([]string, error) {
+	client := g.getClient()
+
+	// get all the releases
+	// NB. currently Github API does not support result ordering, so it not possible to limit results
+	releases, _, err := client.Repositories.ListReleases(context.Background(), g.owner, g.repository, nil)
+	if err != nil {
+		return nil, g.handleGithubErr(err, "failed to get the list of releases")
+	}
+	var versions []string //nolint
```

Comment:
nolint here because we don't know the length in advance.

Reply:
You can try `versions := []string{}`

Reply:
Thanks! Done